provider/core/src/response.rs

```diff
@@ -172,6 +175,74 @@ impl From<&[u8]> for RcWrap<[u8]> {
     }
 }
 
+#[cfg(feature = "sync")]
+impl RcWrap<dyn Any + Send + Sync> {
+    /// Attempt to downcast the `RcWrap<dyn Any + Send + Sync>` to a concrete type.
+    pub fn downcast<T: Any + Send + Sync>(
+        self,
+    ) -> Result<RcWrap<T>, RcWrap<dyn Any + Send + Sync>> {
+        match self.0.downcast() {
+            Ok(s) => Ok(RcWrap(s)),
+            Err(e) => Err(RcWrap(e)),
+        }
+    }
+}
+
+#[cfg(not(feature = "sync"))]
+impl RcWrap<dyn Any> {
+    /// Attempt to downcast the `RcWrap<dyn Any>` to a concrete type.
+    pub fn downcast<T: Any>(self) -> Result<RcWrap<T>, RcWrap<dyn Any>> {
+        match self.0.downcast() {
+            Ok(s) => Ok(RcWrap(s)),
+            Err(e) => Err(RcWrap(e)),
+        }
+    }
+}
+
+#[cfg(feature = "sync")]
+impl<T: Send + Sync + 'static> RcWrap<T> {
+    /// Creates an `RcWrap<dyn Any + Send + Sync>` from an `RcWrap<T>`.
+    pub fn into_dyn_any(self) -> RcWrap<dyn Any + Send + Sync> {
+        RcWrap(self.0)
+    }
+
+    /// Returns the inner value, if the `RcWrap` has exactly one strong reference.
+    pub fn try_unwrap(self) -> Result<T, Self> {
+        match alloc::sync::Arc::try_unwrap(self.0) {
+            Ok(t) => Ok(t),
+            Err(e) => Err(RcWrap(e)),
+        }
+    }
+}
+
+#[cfg(not(feature = "sync"))]
+impl<T: 'static> RcWrap<T> {
+    /// Creates an `RcWrap<dyn Any>` from an `RcWrap<T>`.
+    pub fn into_dyn_any(self) -> RcWrap<dyn Any> {
+        RcWrap(self.0)
+    }
+
+    /// Returns the inner value, if the `RcWrap` has exactly one strong reference.
+    pub fn try_unwrap(self) -> Result<T, Self> {
+        match alloc::rc::Rc::try_unwrap(self.0) {
+            Ok(t) => Ok(t),
+            Err(e) => Err(RcWrap(e)),
+        }
+    }
+}
+
+cfg_if! {
+    if #[cfg(feature = "sync")] {
+        /// A trait that allows to specify bounds for RcWrap value when the "sync" feature is on.
+        pub trait RcWrapBounds: Send + Sync {}
+        impl<T: Send + Sync> RcWrapBounds for T {}
+    } else {
+        /// A trait that allows to specify bounds for RcWrap value when the "sync" feature is off.
+        pub trait RcWrapBounds {}
+        impl<T> RcWrapBounds for T {}
+    }
+}
```

Comment:
Suggestion: This should be possible without the `cfg_if!`, which introduces a new dependency. Just put the cfg on each line (the `pub trait` and the `impl`).

This would be easier if we had https://github.com/rust-lang/rust/issues/41517 but it's still not stabilized

Reply:
Done.